### ISSUE TITLE
Proposal: PowerShellGet on 'Package Management Light'

### ISSUE THREAD

**Issue by edyoung** [PowerShell/PowerShellGetv2#475]
NB: This is a tentative proposal, not a committed plan. In particular, there is no timeframe for implementation.  Comments & concerns are invited.

Currently, Install-Module calls Install-Package in the PackageManagement module, which calls into the PowerShellGet PackageProvider, which calls the Nuget packageprovider. This convoluted flow makes a lot of things more difficult. This is an outline of how we could try to simplify things.

1) Create a new PowerShell/PackageManagementLight module. This would start as a clone of PackageManagement and would be a subset of the Package Management functionality. It would only support retrieving Nuget packages and PowerShell modules from a Nuget feed (including a directory). There would be no option to install or import package providers or update them outside the PMLight module. All *PackageProvider* cmdlets would be gone, and there would be no Msi, Msu, Chocolatey, etc providers. There would be no dynamic parameters, dynamic loading of providers, reflection.emit logic. Code would be merged from the NugetLightProvider submodule into the main PMLight one. 

TBD: Do we still expose PowerShellGet and Nuget as separate providers?

The cmdlets would have different names than those in packagemanagement, probably install-pkg, register-pkgsource, etc. 
Enable-PackageAliases would create aliases that repoint install-package to install-pkg.

Install-Pkg from a Nuget feed would still work, and code which used PackageManagement just to do this kind of thing should work unchanged if it imports PackageManagementLight and runs Enable-PackageAliases instead. This is useful if we believe Install-Package is useful outside of PowerShellGet to manage Nuget packages. If not, we could make PackageManagementLight a nested module of PowerShellGet with a private interface rather than distribute it separately.

2) Update PowerShellGet to reference PackageManagementLight instead. There might be a temporary mode where it can conditionally use one or the other, for testing and evaluation purposes.

3) Remove the PowerShellGet package provider logic and bake it into PackageManagementLight as C# code. This removes the 'nested runspace' issues that we encounter today and a whole bunch of additional code in both PackageManagement and PowerShellGet.

TBD: how long do we keep the provider code inside PowerShellGet in order to support packagemanagement?

TBD: What happens to the DSC resources?

4) Formally deprecate PackageManagement. Exact process for this is TBD but should follow the process for deprecating a Windows component.

5) Further options beyond this point:
- use the Nuget client libraries in PMLight implementation instead of our own reimplementation
- use PMLight for publishing logic as well rather than calling dotnet/nuget.exe
- remove various bootstrapping steps. Need to work out how this would work with Windows.

**Comment by edyoung** [comment]
@SteveL-MSFT @SydneyhSmith @alerickson

**Comment by SteveL-MSFT** [comment]
I know that OneGet did not become what it was originally envisioned, but do we know if anyone has written other package providers?  If simplifying the code means we can have improvements to PSGet, I'm all for it as long as it supports installing not only modules, but any nupkg (think of scripts that call .NET APIs directly into an assembly from nuget.org).  Being able to use Nuget client libraries (presumably they are available cross platform) instead of nuget.exe would be a big win.

**Comment by edyoung** [comment]
Yes, other people have written package providers (Find-PackageProvider to see them). These would not be supported with PMLight - anyone who wants them would have to use them with the legacy PackageManagement module. They will work for a while but we would effectively be losing investments made in those providers. 

PMLight would support fetching nuget packages from nuget.org and other nuget feeds.

**Comment by edyoung** [comment]
@joeyaiello FYI

**Comment by SteveL-MSFT** [comment]
@edyoung the list of providers I see are all from PowerShell team members.  Seems like we wouldn't be losing much to deprecate PkgMgmt so we can put those resources solely towards PSGet

**Comment by ThomasNieto** [comment]
@SteveL-MSFT there are other package management providers created by non-PowerShell team members. The most used package providers were created by the PS team: PSGet at 7M downloads, ChocolateyGet at 340K, Docker at 134K, ContainerImage at 76K. In fact, I created a provider for the packaged software my company produces last month.

Package Management is a highly used module that solves a very real need in the PowerShell community. The software landscape is always changing and there is a new packaging system it seems on a monthly basis. Package Management was created to give users a single set of commands to interface with any packaging system. This need is even greater since PowerShell has become available on Linux and Mac.

It has never hit critical mass due to the lack of providers and support for existing providers. Package Management never had a lot of other package providers because it seems like it was never fully implemented. For example, the Chocolatey team never completed their provider due to the lack of `Update-Package` cmdlet. Its been outstanding since 2014 and is the oldest open issue [#6](https://github.com/OneGet/oneget/issues/6) for OneGet. PowerShellGet did their own implementation (`Update-Module`) instead of adding that API into Package Management. The core providers do not fully support all the package management APIs, for example the Programs provider only supports `Get-Package`.

Here a few of the ways I use Package Management:

- NuGet provider to keep PowerShell gallery and my internal repository in sync.
- Core providers: Programs, MSI, and MSU providers to manage and inventory packages across the enterprise. 
- ChocolateyGet to manage Chocolatey packages
- Package Management DSC resources to install PS Modules, PS scripts, Chocolatey, and other packages
- My company's provider to manage the internally created packages.

I believe that Package Management was a good idea to solve the reality that we live in where there will always be multiple packaging systems. However, that dream was never fully realized due to core features never getting implemented. Package Management can be saved and turned into the thing that it was always meant to be but the core outstanding issues need to be resolved.

Ever since the PowerShell Summit this year I've been wondering how I would to give back to the community. I would like make some major contributions to PowerShellGet and PackageManagment to address the communities most frustrating issues and most wanted enhancements.

**Comment by edyoung** [comment]
Thanks for your input Thomas! Great to hear from people who are getting value from PackageManagement, and of course enhancements and improvements are always welcome.

One reason I am personally a bit less enthusiastic about PackageManagement is that it's actually very difficult to provide an abstraction over many kinds of package which behaves consistently and covers enough use cases. We wind up with issues like  https://github.com/OneGet/oneget/issues/435 , https://github.com/OneGet/oneget/issues/380 , https://github.com/OneGet/oneget/issues/273 where the differences between the underlying systems 'poke through'. It's not clear that end-users can forget about where a package comes from or how it is packaged and just rely on our abstraction. We get into questions like "what should update-package do for an installed program?" 

You could see this issue as a vote in favor of doing a smaller set of things (nuget packages and powershell modules) really well, and trimming the things we don't do as successfully to simplify the code and reduce the surface area.

A possible alternative direction would be for PackageManagement to evolve into an awesome, fully community-owned tool.

**Comment by ThomasNieto** [comment]
@edyoung Thanks for getting feedback from the community.

I'm glad you mentioned these issues, which are a few of them that I actually wanted to bring up myself because I have some recommendations/ideas that I can offer.

- [OneGet/oneget#435](https://github.com/OneGet/oneget/issues/435) and [OneGet/oneget#273](https://github.com/OneGet/oneget/issues/273) are what I refer to in my previous post about incomplete package providers. The `Programs` provider only implemented `Get-Package`. I would say this is more of a provider implementation issue rather than Package Management. My recommendation would be to implement APIs that make sense for the provider and update Package Management to throw an error for non-implemented APIs.

- [OneGet/oneget#380](https://github.com/OneGet/oneget/issues/380) is a provider issue not Package Management. You would still have this issue if it was a stand-alone module querying for MSU.

You can refer back to [OneGet/oneget#6](https://github.com/OneGet/oneget/issues/6) for the full discussion on adding support for `Update-Package`. To condense it, Package Management's role is to provide a flexible framework for providers to build upon. It is up to the provider to determine which APIs to implement. It may not make sense for the `Programs` provider to support `Update-Package`, `*-PackageSource`, or `Find-Package`. By updating Package Management to write an error for non-implemented APIs, the end user receives clear feedback. End user confusion comes into play when the cmdlet does not write an error giving the user the expectation that it worked. The following example illustrates this scenario:

### Current
```powershell
C:\> Get-Package -Name *notepad*

Name                           Version          Source                           ProviderName
----                           -------          ------                           ------------
Notepad++ (64-bit x64)         7.6.6                                             Programs


C:\> Get-Package -Name *notepad* | Uninstall-Package
C:\>
```

### Ideal
```powershell
C:\> Get-Package -Name *notepad*

Name                           Version          Source                           ProviderName
----                           -------          ------                           ------------
Notepad++ (64-bit x64)         7.6.6                                             Programs


C:\> Get-Package -Name *notepad* | Uninstall-Package
The 'Programs' provider does not support 'Uninstall-Package'.
    + CategoryInfo          : InvalidOperation: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : ProviderOperationNotSupported
```

I would disagree that doing smaller number of things well (NuGet and PS Modules) is mutually exclusive with the current landscape. I would say that the implementing individual package providers accomplishes exactly that, isolating the code base to a single package system. During the last year how many issues would you attribute to package providers (NuGet, PSGet, etc) compared to Package Management core?

Correct me if I'm wrong but supporting the smaller number of things would be removing support for the core providers: `Programs`, `MSI`, and `MSU` and third party providers. In the past year how many issues were attributed to those three core providers?

The PowerShell team has indicated that they would like to create a provider (plug-in) model for credentials and logging/tracing. These types of issues will still be present in those projects. Also, if we're having issues supporting Package Management framework and three core providers, I have concerns for the long term viability of those new projects.

In summary, PowerShell providers (FileSystem, Registry, etc) have been used and understood by the community for many years. Not every PowerShell provider supports all the available commands (*-Item) but will provide end user feedback if the operation performed is not supported. Users then learn through documentation and trial and error to understand the capabilities of each provider. Package Management serves a similar role for packaging systems. The problem today is that the feedback loop does not exist in the current cmdlets misleading users to think the cmdlet worked as expected. Once that enhancement is made, package providers begin to make sense to users.

**Comment by alerickson** [comment]
Is there an advantage to creating a PackageManagement Light vs integrating the Nuget v3 API?  It seems like the latter would be much less work and still separate out PowerShellGet from PackageManagement so that they can be developed independently. Potentially PackageManagement Light could use Nuget API calls too.

**Comment by edyoung** [comment]
@tnieto88 I don't think this says anything about the benefits of a provider/plugin model elsewhere in PowerShell - what are PowerShell cmdlets after all but plugins? I'm also not really making any claim about whether bugs are in the providers or package management itself.

When it gets right down to it, my concerns are much more basic. The code required to Find-Module is about 5x as much as it would be if it just directly implemented fetching a module from a nuget repo without this intermediate abstraction, and is quite difficult to debug and update. An example of how this hurts us is #34. This looks trivial but is surprisingly complicated to fix because the prompt occurs in a nested PowerShell runspace spawned inside PackageManagement. That's something we just wouldn't do otherwise.

Managing PowerShell modules is far and away the most popular use of PackageManagement, and it's the only one I know of that has no reasonable alternative - I think every other kind of package handled by Package Management has at least a command-line interface if not a first-class PowerShell interface. 

An alternative direction  is similar to what @alerickson suggests:

1) implement a minimal nuget fetching library as a nested module inside PowerShellGet, cutting its dependency on Package Management. This wouldn't be public so we wouldn't have an Install-Pkg.

2) Once the critical path no longer depends on it, remove PackageManagement from the PowerShell distribution. It's just an Install-Module away and wouldn't lose functionality (although we probably want to remove the PowerShellGet provider).

3) Set PackageManagment free to be enhanced by the community. I'm not a big fan of creating what could reasonably be called "abandonware' but it seems better than just deleting it.

Downsides are:
- we'd no longer have PackageManagement functionality in-box
- PM has some dependencies on services run by Microsoft (downloading files from oneget.org) which might take some unravelling
- getting the updated PowerShellGet to be exactly compatible with the existing one would be very challenging

**Comment by ThomasNieto** [comment]
@edyoung Let me clarify, my concern is with the breaking change of removing the `PowerShellGet` provider and `PackageManagement` module, not where the internal logic is located. I completely understand and agree with you that the current implementation having the logic in the `PowerShellGet` provider causes more problems than it solves.

I also agree with @alerickson that the best approach would be to implement the NuGet API directly in `PowerShellGet` module removing the dependency on `PackageManagement` allowing them to be developed independently.

To continue supporting the use case of managing modules and scripts from `PackageManagement`, the `PowerShellGet` provider would need to be updated. The work effort would be minimal as the only logic in the provider would be calling the `PowerShellGet` methods/functions and the required logic for the provider to function. This allows users to continue to have dedicated PowerShell cmdlets (`*-Module`, `*-Script`) as well as supporting `PackageManagement` cmdlets.

I believe this would solve the two main concerns you have with `PackageManagement`: nested runspace and extra complexity of having logic split between the provider and module. This is the approach I've gone with for my company's provider with great success. It's also the recommended approach for creating DSC resources. By having all the logic in the PowerShell module, the DSC resource will only have the required logic to support DSC.

@joeyaiello has PowerShell ever removed an in-box module outside the technical limitations of the 5.1 to 6.0 release? If so does that constitute a breaking change as the cmdlets are no longer present by default?

**Comment by edyoung** [comment]
Please see the more extensive proposal referenced above, we won't be proceeding with this.
